waku/v2/protocol/waku_filter/client.nim

```diff
@@ -1,69 +1,207 @@
+  SubscriptionManager(subscriptions: newTable[(string, ContentTopic), FilterPushHandler]())
+
+proc clear(m: var SubscriptionManager) =
+  m.subscriptions.clear()
+
+proc registerSubscription(m: SubscriptionManager, pubsubTopic: string, contentTopic: ContentTopic, handler: FilterPushHandler) =
+  try:
+    m.subscriptions[(pubsubTopic, contentTopic)]= handler
+  except:
+    error "failed to register filter subscription", error=getCurrentExceptionMsg()
+
+proc removeSubscription(m: SubscriptionManager, pubsubTopic: string, contentTopic: ContentTopic) =
+  m.subscriptions.del((pubsubTopic, contentTopic))
+
+proc notifySubscriptionHandler(m: SubscriptionManager, pubsubTopic: string, contentTopic: ContentTopic, message: WakuMessage) =
+  if not m.subscriptions.hasKey((pubsubTopic, contentTopic)):
+    return
+
+  try:
+    let handler = m.subscriptions[(pubsubTopic, contentTopic)]
+    handler(pubsubTopic, message)
+  except:
+    discard
+
+proc getSubscriptionsCount(m: SubscriptionManager): int =
+  m.subscriptions.len()
+
+
+## Client
+
+type  MessagePushHandler* = proc(requestId: string, msg: MessagePush): Future[void] {.gcsafe, closure.}
+
+type WakuFilterClient* = ref object of LPProtocol
+    rng: ref rand.HmacDrbgContext
+    peerManager: PeerManager
+    subManager: SubscriptionManager
+
+
+proc handleMessagePush(wf: WakuFilterClient, peerId: PeerId, requestId: string, rpc: MessagePush) =
+  for msg in rpc.messages:
+    let 
+      pubsubTopic = Defaultstring # TODO: Extend the filter push rpc to provide the pubsub topic. This is a limitation
+      contentTopic = msg.contentTopic 
+    
+    wf.subManager.notifySubscriptionHandler(pubsubTopic, contentTopic, msg)
+
+
+proc initProtocolHandler(wf: WakuFilterClient) =
+  proc handle(conn: Connection, proto: string) {.async, gcsafe, closure.} =
+    let buffer = await conn.readLp(MaxRpcSize.int)
+
+    let decodeReqRes = FilterRPC.init(buffer)
+    if decodeReqRes.isErr():
+      waku_filter_errors.inc(labelValues = [decodeRpcFailure])
+      return
+
+    let rpc = decodeReqRes.get()
+    trace "filter message received"
+
+    if rpc.push == MessagePush():
+      waku_filter_errors.inc(labelValues = [emptyMessagePushFailure])
+      # TODO: Manage the empty push message error. Perform any action?
+      return
+
+    waku_filter_messages.inc(labelValues = ["MessagePush"])
+    
+    let
+      peerId = conn.peerId
+      requestId = rpc.requestId 
+      push = rpc.push
+
+    info "received filter message push", peerId=conn.peerId, requestId=requestId
+    wf.handleMessagePush(peerId, requestId, push)
+
+  wf.handler = handle
+  wf.codec = WakuFilterCodec
+
+proc new*(T: type WakuFilterClient,
+          peerManager: PeerManager,
+          rng: ref rand.HmacDrbgContext): T = 
+          
+  let wf = WakuFilterClient(
+      peerManager: peerManager,
+      rng: rng,
+      subManager: SubscriptionManager.init()
+    )
+  wf.initProtocolHandler()
+  wf
```

Comment:
i would enforce using `return` in every function. thoughts?

Reply:
What is the rationale behind the rule you are suggesting?

Reply:
1) consistency and 2) readability.

i.e. we use `return ok()` but then `wf`.

do you have any strong reason for avoiding the `return` here?

Reply:
We use `return ok()` because we cannot do just `ok()`. But this is only true if the is annotated with `{.async.}` pragma and returns a result. In that case, the compilation fails, and it refuses to compile. So the only solution is to do a `return ok()`.

As reference: https://status-im.github.io/nim-style-guide/language.result.html#practical-notes

Reply:
okay great input. I only tried **without** {.async.} and `ok()` works, but yeah not with {.async.}, where `return` is required.

But this kind of backs the decision of enforcing `return` since in some cases without it, it doesn't work (like ok() inside {.async.}